### ISSUE TITLE
Error Loading pyRevit: Error dynamically compiling pyRevit runtime

### ISSUE THREAD

**Issue by lumpentrunk** [eirannejad/pyRevit#1948]
**Describe the bug**
Hi guys, hopefully someone can help me. 
I'm using Revit 21 and Revit 22 with no admin rights.

The installation of pyRevit completed with no errors but pyRevit didn't load, so I ran the command below in terminal. 

`pyrevit attach master 277 --installed`

But it didn't solved the Problem at all. A new error message appears on every start of Revit.

**Screenshots**
![image](https://github.com/eirannejad/pyRevit/assets/13121268/a943fb91-df94-4fbb-8923-d52ffafbf9c0)
![image](https://github.com/eirannejad/pyRevit/assets/13121268/f2b63dea-d0e8-47da-9e9a-cc24791c4952)

**Desktop (please complete the following information):**
 - OS: Win10
 - pyRevit Version: 4.8.13.23182 (installed with the non-admin installer pyRevit_4.8.13.23182_signed.exe)
 - pyRevit Environment: 
 
 ```
C:\Users\xxx>pyrevit env
==> Registered Clones (full git repos)
==> Registered Clones (deployed from archive/image)
master | Deploy: "basepublic" | Branch: "master" | Version: "4.8.13.23182+2215" | Path: "C:\Users\xxx\AppData\Roaming\pyRevit-Master"
==> Attachments
master | Product: "2022.1.3" | Engine: IPY277 (277) | Path: "C:\Users\xxx\AppData\Roaming\pyRevit-Master"
master | Product: "2021.1.6" | Engine: IPY277 (277) | Path: "C:\Users\xxx\AppData\Roaming\pyRevit-Master"
==> Installed Extensions
==> Default Extension Search Path
C:\Users\xxx\AppData\Roaming\pyRevit\Extensions
==> Extension Search Paths
==> Extension Sources - Default
https://github.com/eirannejad/pyRevit/raw/master/extensions/extensions.json
==> Extension Sources - Additional
==> Installed Revits
2022.1.3 | Version: 22.1.30.34 | Build: 20220520_1515(x64) | Language: 1033 | Path: "C:\Program Files\Autodesk\Revit 2022\"
2021.1.6 | Version: 21.1.60.25 | Build: 20220123_1515(x64) | Language: 1033 | Path: "C:\Program Files\Autodesk\Revit 2021\"
==> Running Revit Instances
PID: 29788 | 2022.1.3 | Version: 22.1.30.34 | Build: 20220520_1515(x64) | Language: 0 | Path: "C:\Program Files\Autodesk\Revit 2022"
==> User Environment
Microsoft Windows 10 [Version 10.0.19045]
Executing User: yyy\xxx
Active User: yyy\xxx
Admin Access: No
%APPDATA%: "C:\Users\xxx\AppData\Roaming"
Latest Installed .Net Framework: 4.8
No .Net Target Packs are installed.
No .Ne-Core Target Packs are installed.
pyRevit CLI v4.8.13.23182+2215
```

Thanks for your support!

**Comment by jmcouffin** [comment]
try this one `pyrevit attach master 2711 --installed --allusers` and a good restart.

**Comment by lumpentrunk** [comment]
thx for the tip, but it doesn't work. I get another error message.

I did the following:
I opened the cmd window and typed in
```pyrevit attach master 2711 --installed --allusers```

after this, for the "good restart", I typed in ```shutdown /g /f /t 0``` in the run dialog (```win+R```)

after the restart I started revit 22 and now I get the following error message:
![image](https://github.com/eirannejad/pyRevit/assets/13121268/3f2621c0-225d-4231-85d0-f92871282b9b)
![image](https://github.com/eirannejad/pyRevit/assets/13121268/1997ec1c-a873-4c4f-9c75-140a253f88d2)

**Comment by lumpentrunk** [comment]
```
C:\Users\xxx>pyrevit env
==> Registered Clones (full git repos)
==> Registered Clones (deployed from archive/image)
master | Deploy: "basepublic" | Branch: "master" | Version: "4.8.13.23182+2215" | Path: "C:\Users\xxx\AppData\Roaming\pyRevit-Master"
==> Attachments
master | Product: "2022.1.3" | Engine: DEFAULT (2711) | Path: "C:\Users\xxx\AppData\Roaming\pyRevit-Master" | AllUsers
master | Product: "2021.1.6" | Engine: DEFAULT (2711) | Path: "C:\Users\xxx\AppData\Roaming\pyRevit-Master" | AllUsers
==> Installed Extensions
==> Default Extension Search Path
C:\Users\xxx\AppData\Roaming\pyRevit\Extensions
==> Extension Search Paths
==> Extension Sources - Default
https://github.com/eirannejad/pyRevit/raw/master/extensions/extensions.json
==> Extension Sources - Additional
==> Installed Revits
2022.1.3 | Version: 22.1.30.34 | Build: 20220520_1515(x64) | Language: 1033 | Path: "C:\Program Files\Autodesk\Revit 2022\"
2021.1.6 | Version: 21.1.60.25 | Build: 20220123_1515(x64) | Language: 1033 | Path: "C:\Program Files\Autodesk\Revit 2021\"
==> Running Revit Instances
PID: 15516 | 2022.1.3 | Version: 22.1.30.34 | Build: 20220520_1515(x64) | Language: 0 | Path: "C:\Program Files\Autodesk\Revit 2022"
==> User Environment
Microsoft Windows 10 [Version 10.0.19045]
Executing User: yyy\xxx
Active User: yyy\xxx
Admin Access: No
%APPDATA%: "C:\Users\xxx\AppData\Roaming"
Latest Installed .Net Framework: 4.8
No .Net Target Packs are installed.
No .Ne-Core Target Packs are installed.
pyRevit CLI v4.8.13.23182+2215
```

**Comment by lumpentrunk** [comment]
Good morning,
today I get yesterdays first error message again "Error dynamically compiling pyRevit runtime". Very strange ... I just shuted down my computer yesterday and started it today ... 

![image](https://github.com/eirannejad/pyRevit/assets/13121268/6b8782b1-c6dc-4cf0-9936-54623d6e107b)

**Comment by jmcouffin** [comment]
> Good morning, today I get yesterdays first error message again "Error dynamically compiling pyRevit runtime". Very strange ... I just shuted down my computer yesterday and started it today ...
> 
> ![image](https://user-images.githubusercontent.com/13121268/276866984-6b8782b1-c6dc-4cf0-9936-54623d6e107b.png)

I really have no clue on that one.

what I would try:

- clean uninstall - install latest
- deactivating all revit addins
- check if an antivirus or firewall block something (sentinelOne again?)

**Comment by jmcouffin** [comment]
https://github.com/eirannejad/pyRevit/issues/1989#issuecomment-1821102231

also 
Please try one of the installers in the artifacts section of this page

https://github.com/eirannejad/pyRevit/actions/runs/6962280865

These are WIP installers and implements changes that could fix the issues

note: uninstall first...

**Comment by bradbd42** [comment]
Was also having this same issue. It is working now after performing the below steps.

1. Uninstalled pyRevit
2. Uninstalled Revit
3. Deleted ProgramData and AppData pyRevit folders
4. Moved all Revit addins (ProgramData\Revit\Addins) to a temporary location
5. Rebooted computer
6. Reinstalled Revit
7. Reinstalled pyRevit
8. Confirmed Revit could be opened without issue
9. Restored addins moved in step 4

**Comment by jmcouffin** [comment]
Thanks for the feedback @bradbd42 

Not sure this step was necessary, but I am glad it worked for you
> 2. Uninstalled Revit

**Comment by jmcouffin** [comment]
@lumpentrunk fixed?

**Comment by BethanyBIM** [comment]
I was having the same problem and can confirm that uninstalling Revit is NOT necessary. I struggled a bit with the steps provided above, especially because I have several versions of Revit installed, but finally figured it out so here is some clarification on the steps that worked here:

1. Uninstall pyRevit
2. Delete C:\ProgramData and C:\Users\%username%\AppData\Roaming pyRevit folders
3. Moved all Revit addins (C:\ProgramData\Autodesk\Revit\Addins) to a temporary location
4. Rebooted computer
5. Reinstalled pyRevit
6. Started **each version** of Revit confirming pyRevit successfully compiled at startup
7. Moved all Revit addins that were moved in step 3 back to original location

**Comment by jmcouffin** [comment]
Let me close this one, as the original poster did not react to others solutions.
Feel free to reopen if this makes sense.

I added your steps to the pyRevit documentation with some minor corrections @BethanyBIM :

> 1. **Uninstall pyRevit**
> 2. **Delete** C:\ProgramData and \%AppData%\ pyRevit folders
> 3. **Move all Revit addins** to a temporary location. Usual addins locations include:
>     1. %AppData%\Autodesk\Revit\Addins\<year>
>     2. %AppData%\Autodesk\ApplicationPlugins
>     3. %ProgramData%\Autodesk\Revit\Addins\<year>
>     4. %ProgramData%\Autodesk\ApplicationPlugins
> 4. [Reboot computer]
> 5. **Reinstall pyRevit**
> 6. Started **each version** of Revit confirming pyRevit successfully compiled at startup.
> 7. **Moved all Revit addin**s that were moved in step 3 back to their original location.
>     1. you may want to add them one at a time and start a revit session everytime to figure out which ones are causing the issue.

https://www.notion.so/pyrevitlabs/Installation-issues-f6495adb56254e9e8021f8a68e3b2ab7?pvs=4#ee9590e3dd8d4920b8b09ecb3c42d993

Thanks for your contributions

**Comment by LeslieAtWork** [comment]
We're running into this issue on many computers and doing the fix above works, but it takes about 20 minutes per machine. Is there any information on the root cause of this? If there's no better fix, our solution will probably have to be to never update pyRevit again.

**Comment by jmcouffin** [comment]
> the root cause of this
@LeslieAtWork 
Dll hell and the loading sequence of dll in Revit.
Revit loads addins and their dlls and referenced dlls in a sequence.
The referenced ones are stacking and if one is the same as for another dlls, then, one as precedence over the other. One of the 2 addins is then missing a reference to work properly.
We expect to get this fixed with the next version of pyRevit. But this is a tough one.

**Comment by LeslieAtWork** [comment]
> > We expect to get this fixed with the next version of pyRevit. But this is a tough one.
Thank you for the quick reply. It's good to know it's on the radar.

**Comment by aoffak** [comment]
> > the root cause of this
> > @LeslieAtWork
> > Dll hell and the loading sequence of dll in Revit.
> > Revit loads addins and their dlls and referenced dlls in a sequence.
> > The referenced ones are stacking and if one is the same as for another dlls, then, one as precedence over the other. One of the 2 addins is then missing a reference to work properly.
> > We expect to get this fixed with the next version of pyRevit. But this is a tough one.

Has anyone tried to see if updating IronPython fixed this issue per Autodesk's suggestion? https://www.autodesk.com/support/technical/article/caas/sfdcarticles/sfdcarticles/PyRevitLoader-Error-Loading-pyRevit-message-when-opening-Revit.html

**Comment by jmcouffin** [comment]
> Has anyone tried to see if updating IronPython fixed this issue per Autodesk's suggestion? https://www.autodesk.com/support/technical/article/caas/sfdcarticles/sfdcarticles/PyRevitLoader-Error-Loading-pyRevit-message-when-opening-Revit.html

Good luck with that one!
This article is totally missing the point.
Why would you install iron python 2.7.11 when pyrevit comes shipped with the 2.7.11 engine already?

The issue is the regular DLL hell one. nothing more.
